[Merged by Bors] - feat(linear_algebra/matrix): Spectral theorem for matrices

abentkamp opened the pull request:
---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

abentkamp committed:
hermitian matrices

abentkamp committed:
formulation of spectral thm

abentkamp committed:
more spectral

abentkamp committed:
spectral theorem for matrices

abentkamp committed:
add comments and move lemmas

abentkamp committed:
add file header

eric-wieser commented on src/analysis/inner_product_space/matrix_spectrum.lean:
```diff
@@ -0,0 +1,108 @@
+
+/-- A matrix is hermitian if it is equal to its conjugate transpose. On the reals, this definition
+captures symmetric matrices. -/
+def is_hermitian (A : matrix n n 𝕜) : Prop := Aᴴ = A
```
This probably belongs next to `linear_algebra/matrix/symmetric`; or even in the same file.

abentkamp replied:
The spectral theorem should probably also be in `linear_algebra/matrix`, right? I've moved the whole file there and split it into `spectrum.lean` and `hermitian.lean`.

abentkamp committed:
fix inner_eq_star_dot_product

abentkamp committed:
fix Type*

abentkamp committed:
move file

abentkamp committed:
fix proof

abentkamp committed:
put hermitian in separate file

eric-wieser commented on src/analysis/inner_product_space/pi_L2.lean:
```diff
@@ -113,6 +113,10 @@ instance : inner_product_space 𝕜 (euclidean_space 𝕜 ι) := by apply_instan
   finite_dimensional.finrank 𝕜 (euclidean_space 𝕜 (fin n)) = n := by simp
 
+lemma euclidean_space.inner_eq_star_dot_product (x y : euclidean_space 𝕜 ι) :
+  ⟪x, y⟫ = matrix.dot_product (star x) y :=
```
I'd argue this should be spelt
```suggestion
  ⟪x, y⟫ = matrix.dot_product (star $ pi_Lp.equiv _ _ x) (pi_Lp.equiv _ _ y) := rfl
```
since that avoids unfolding type synonyms in surprising ways.

I think @hrmacbeth will probably disagree with me though.

abentkamp replied:
I don't have an opinion on this, but I've implemented your suggestion.

eric-wieser commented on src/data/matrix/basic.lean:
```diff
@@ -1480,6 +1480,20 @@ instance [fintype n] [semiring α] [star_ring α] : star_ring (matrix n n α) :=
+  ... = (star M).vec_mul (star v) i :
+    by simp only [vec_mul, (star_apply _ _ _).symm, conj_transpose_apply]
+end
+
```
I've split this out into #14248, since I was able to golf the proof down to 1 line and generalize the typeclass arguments.

abentkamp replied:
Oh, wow, that's great!

eric-wieser replied:
A good strategy for proving trivial things about `vec_mul` and `mul_vec` is just to unfold everything, and then see if what you end up with is something you can already prove about `dot_product`.

eric-wieser replied:
I pushed a merge commit that removed this lemma

abentkamp committed:
fix linting errors

abentkamp committed:
fix more linting errors

abentkamp committed:
fix renaming

abentkamp committed:
use pi_Lp.equiv

abentkamp committed:
fix linting error

eric-wieser committed:
Merge remote-tracking branch 'origin/master' into matrix_spectrum

Vierkantor commented on src/analysis/inner_product_space/pi_L2.lean:
```diff
@@ -470,4 +473,18 @@ lemma inner_matrix_row_row (A B : matrix (fin n) (fin m) 𝕜) (i j : (fin n)) :
+matrix.mul_vec_std_basis A j i
+
+@[simp] lemma euclidean_space.mul_vec_single [decidable_eq ι] (A : matrix ι' ι 𝕜) :
+  A.mul_vec (euclidean_space.single i 1) = λ j, A j i :=
```
Below you use the transpose instead, so should this not be:
```suggestion
  A.mul_vec (euclidean_space.single i 1) = Aᵀ j :=
```

Vierkantor commented on src/linear_algebra/matrix/basis.lean:
```diff
@@ -177,6 +177,33 @@ lemma basis_to_matrix_mul_linear_map_to_matrix_mul_basis_to_matrix
+  simp only [basis_to_matrix_mul _ _ (pi.basis_fun R ι), matrix.to_lin_eq_to_lin'],
+  ext i j,
+  have : A.mul_vec ((linear_map.std_basis R (λ (i : ι), R) j) 1) = Aᵀ j :=
+    funext (λ x, matrix.mul_vec_std_basis A x j),
```
Should this become its own lemma to mirror the situation around `euclidean_space.mul_vec_single`?

Vierkantor commented on src/linear_algebra/matrix/hermitian.lean:
```diff
@@ -0,0 +1,53 @@
+  { intro h,
+    ext i j,
+    have := h (euclidean_space.single i 1) (euclidean_space.single j 1),
+    simpa [euclidean_space.inner_single_right, euclidean_space.inner_single_left] using this}
```
Weirdly, `simpa using` takes a whole expression while `rwa at` only takes a variable name.
```suggestion
    simpa [euclidean_space.inner_single_right, euclidean_space.inner_single_left]
      using h (euclidean_space.single i 1) (euclidean_space.single j 1) }
```

Vierkantor commented on src/linear_algebra/matrix/spectrum.lean:
```diff
@@ -0,0 +1,87 @@
+
+/-- The eigenvalues of a hermitian matrix, reusing the index `n` of the matrix entries. -/
+noncomputable def eigenvalues : n → ℝ :=
+  λ i, hA.eigenvalues₀ $ (fintype.equiv_of_card_eq (fintype.card_fin _)).symm i
```
```suggestion
λ i, hA.eigenvalues₀ $ (fintype.equiv_of_card_eq (fintype.card_fin _)).symm i
```

Vierkantor commented on src/linear_algebra/matrix/spectrum.lean:
```diff
@@ -0,0 +1,87 @@
+noncomputable def eigenvector_basis : basis n 𝕜 (n → 𝕜) :=
+  (@inner_product_space.is_self_adjoint.eigenvector_basis 𝕜 _ _
+  (pi_Lp 2 (λ (_ : n), 𝕜)) _ A.to_lin' (is_hermitian_iff_is_self_adjoint.1 hA) _ (fintype.card n)
+  finrank_euclidean_space).reindex (fintype.equiv_of_card_eq (fintype.card_fin _))
```
```suggestion
(@inner_product_space.is_self_adjoint.eigenvector_basis 𝕜 _ _
    (pi_Lp 2 (λ (_ : n), 𝕜)) _ A.to_lin' (is_hermitian_iff_is_self_adjoint.1 hA) _
    (fintype.card n) finrank_euclidean_space).reindex
  (fintype.equiv_of_card_eq (fintype.card_fin _))
```

Vierkantor replied:
All these `fintype.equiv_of_card_eq` suggest that it's worth generalizing the definitions about eigenvectors/-values in `inner_product_space` from `fin (finrank k E)` to `ι`, where `[fintype ι] (h : basis ι k E)`. And then define a new `sorted_eigenvalues` that lists them in increasing order as a `fin (finrank k E) → ℝ` tuple (as the TODO notes).

What do you think, would it be worth trying to do this change in a follow-up PR?

Vierkantor commented on src/linear_algebra/matrix/spectrum.lean:
```diff
@@ -0,0 +1,87 @@
+
+/-- A matrix whose columns are an orthonormal basis of eigenvectors of a hermitian matrix. -/
+noncomputable def eigenvector_matrix : matrix n n 𝕜 :=
+  (pi.basis_fun 𝕜 n).to_matrix (eigenvector_basis hA)
```
```suggestion
(pi.basis_fun 𝕜 n).to_matrix (eigenvector_basis hA)
```

Vierkantor commented on src/linear_algebra/matrix/spectrum.lean:
```diff
@@ -0,0 +1,87 @@
+
+/-- The inverse of `eigenvector_matrix` -/
+noncomputable def eigenvector_matrix_inv : matrix n n 𝕜 :=
+  (eigenvector_basis hA).to_matrix (pi.basis_fun 𝕜 n)
```
```suggestion
(eigenvector_basis hA).to_matrix (pi.basis_fun 𝕜 n)
```

Vierkantor commented on src/linear_algebra/matrix/spectrum.lean:
```diff
@@ -0,0 +1,87 @@
+For the spectral theorem on linear maps, see `diagonalization_basis_apply_self_apply`. -/
+theorem spectral_theorem :
+  hA.eigenvector_matrix_inv ⬝ A
+    = diagonal (coe ∘ hA.eigenvalues) ⬝ hA.eigenvector_matrix_inv :=
```
```suggestion
  hA.eigenvector_matrix_inv ⬝ A =
    diagonal (coe ∘ hA.eigenvalues) ⬝ hA.eigenvector_matrix_inv :=
```

Vierkantor commented on src/linear_algebra/matrix/spectrum.lean:
```diff
@@ -0,0 +1,87 @@
+      inner_product_space.is_self_adjoint.diagonalization_basis],
+    rw [basis.to_matrix_apply, basis.coe_to_orthonormal_basis_repr, basis.reindex_repr,
+      basis.equiv_fun_apply, pi.basis_fun_apply],
+    refl }
```
We can remove the `refl` with a little bit of rewriting:
```suggestion
    rw [basis.to_matrix_apply, basis.coe_to_orthonormal_basis_repr, basis.reindex_repr,
      basis.equiv_fun_apply, pi.basis_fun_apply, eigenvalues₀, linear_map.coe_std_basis,
      euclidean_space.single] }
```

Vierkantor commented on src/linear_algebra/matrix/spectrum.lean:
```diff
@@ -0,0 +1,87 @@
+      to_lin'_apply],
+    simp only [basis.to_matrix, basis.coe_to_orthonormal_basis_repr, basis.equiv_fun_apply],
+    rw [basis.reindex_repr, euclidean_space.mul_vec_single],
+    refl },
```
You should be able to get rid of this `refl`  by modifying `mul_vec_single` to use the transpose instead.

Vierkantor reviewed (approved):
bors d+

eric-wieser commented on src/linear_algebra/matrix/hermitian.lean:
```diff
@@ -0,0 +1,53 @@
+
+/-- A matrix is hermitian if it is equal to its conjugate transpose. On the reals, this definition
+captures symmetric matrices. -/
+def is_hermitian (A : matrix n n 𝕜) : Prop := Aᴴ = A
```
This definitionn should be generalized to any `star_ring`, not just `is_R_or_C`

eric-wieser replied:
You should aim to copy most of the API for `matrix.is_symm`

eric-wieser commented on src/linear_algebra/matrix/hermitian.lean:
```diff
@@ -0,0 +1,53 @@
+
+/-- A matrix is hermitian iff the corresponding linear map is self adjoint. -/
+lemma is_hermitian_iff_is_self_adjoint {A : matrix n n 𝕜} :
+  is_hermitian A ↔ @inner_product_space.is_self_adjoint 𝕜 (pi_Lp 2 (λ (_ : n), 𝕜)) _ _ A.to_lin' :=
```
This feels like a nicer way to state this:
```lean
/-- `pi_Lp.equiv` as a linear map. -/
@[simps {fully_applied := ff}]
protected def _root_.pi_Lp.linear_equiv (p : ℝ) [fact_one_le_p : fact (1 ≤ p)]
  {ι} (R) (α : ι → Type*)
  [fintype ι] [normed_field R] [Π i, semi_normed_group (α i)] [Π i, normed_space R (α i)] :
  pi_Lp p α ≃ₗ[R] Π (i : ι), α i :=
{ to_fun := pi_Lp.equiv _ _,
  inv_fun := (pi_Lp.equiv _ _).symm,
  ..linear_equiv.refl _ _}
```
(added in #14380)
```lean
/-- A matrix is hermitian iff the corresponding linear map is self adjoint. -/
lemma is_hermitian_iff_is_self_adjoint {A : matrix n n 𝕜} :
  is_hermitian A ↔ inner_product_space.is_self_adjoint
    ((pi_Lp.linear_equiv 2 𝕜 (λ _ : n, 𝕜)).symm.conj A.to_lin') :=
```
which I can mostly prove as
```lean
begin
  rw [inner_product_space.is_self_adjoint, (pi_Lp.equiv 2 (λ _ : n, 𝕜)).symm.surjective.forall₂],
  simp only [linear_equiv.conj_apply, linear_map.comp_apply, linear_equiv.coe_coe,
    pi_Lp.linear_equiv_apply, pi_Lp.linear_equiv_symm_apply, linear_equiv.symm_symm],
  simp_rw [
    euclidean_space.inner_eq_star_dot_product, equiv.apply_symm_apply, to_lin'_apply,
    star_mul_vec, dot_product_mul_vec],
  split,
  { rintro (h : Aᴴ = A) x y,
    rw h },
  { intro h,
    ext i j,
    have := h (pi.single i 1) (pi.single j 1),
    simp [vec_mul] at this,
    sorry, }
end
```

eric-wieser commented on src/analysis/inner_product_space/pi_L2.lean:
```diff
@@ -470,4 +473,18 @@ lemma inner_matrix_row_row (A B : matrix (fin n) (fin m) 𝕜) (i j : (fin n)) :
+
+lemma euclidean_space.mul_vec_single_apply [decidable_eq ι] (A : matrix ι' ι 𝕜) :
+  A.mul_vec (euclidean_space.single i 1) j = A j i :=
+matrix.mul_vec_std_basis A j i
```
This lemma is a bit suspicious, as it's not type-correct; it ought to have a `pi_Lp.equiv` in there somewhere.

eric-wieser committed:
chore(analysis/normed_space/pi_Lp): add `pi_Lp.linear_equiv`

This is just a more bundled version of the `pi_Lp.equiv` we already have.
Also adds two missing simp lemmas about `pi_Lp.equiv`.

eric-wieser committed:
explicit var

eric-wieser committed:
fix argument order

abentkamp committed:
make lemmas type correct

eric-wieser commented on src/analysis/inner_product_space/pi_L2.lean:
```diff
@@ -470,4 +473,18 @@ lemma inner_matrix_row_row (A B : matrix (fin n) (fin m) 𝕜) (i j : (fin n)) :
+
+@[simp] lemma euclidean_space.vec_mul_single [decidable_eq ι] (A : matrix ι ι' 𝕜) :
+  A.vec_mul (pi_Lp.equiv _ _ (euclidean_space.single i 1)) j = A i j :=
+matrix.vec_mul_std_basis A i j
```
Note that after #14380 these won't be simp-normal form anymore, which will make the linter complain.

Are you happy to wait for that PR to be reviewed / merged before wrapping up this one?

abentkamp replied:
Ok, sure.

eric-wieser replied:
It's now merged

abentkamp committed:
matrix.mul_vec_std_basis_apply

abentkamp committed:
simpa using expr

abentkamp committed:
remove spaces

abentkamp committed:
remove spaces

abentkamp committed:
remove spaces

abentkamp committed:
move eq sign

abentkamp committed:
remove refl

abentkamp committed:
Merge remote-tracking branch 'origin/eric-wieser/pi_Lp.linear_equiv' into matrix_spectrum

abentkamp committed:
undo simpa

abentkamp committed:
fix hermitian proof

abentkamp committed:
simp [-vec_mul_single]

abentkamp committed:
use pi.single instead euclidean_space.single

abentkamp committed:
remove simp attributes

abentkamp committed:
fix matrix diagonalization proof

abentkamp committed:
hermitian api

abentkamp committed:
generalize to star_ring

abentkamp committed:
remove unnecessary lemmas

abentkamp committed:
Update src/linear_algebra/matrix/spectrum.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

abentkamp committed:
Merge branch 'master' into matrix_spectrum

eric-wieser commented on src/data/matrix/basic.lean:
```diff
@@ -1142,6 +1142,10 @@ lemma mul_vec_smul [fintype n] [monoid R] [non_unital_non_assoc_semiring S] [dis
+lemma vec_mul_single [fintype m] [decidable_eq m] (A : matrix m n α) (x : α) (i : m) :
+   vec_mul (pi.single i x) A = x • A i :=
+by { ext, simp [vec_mul, pi.single_apply] }
+
```
Note that a different version of these lemmas already appears in #13835

abentkamp replied:
Hm, yeah, should we wait for #13835 then?

eric-wieser replied:
It's now merged

eric-wieser commented on src/data/matrix/basic.lean:
```diff
@@ -1488,6 +1496,10 @@ lemma star_eq_conj_transpose [has_star α] (M : matrix m m α) : star M = Mᴴ :
 
+lemma star_single [add_monoid α] [star_add_monoid α] [decidable_eq n] (i : n) (a : α) :
+  star (pi.single i a) = pi.single i (star a) :=
+by { ext j, by_cases h : j = i; simp [pi.single_apply, h, star_zero] }
```
This isn't about `matrix`, so doesn't belong in this file

abentkamp committed:
move star_single

abentkamp committed:
Merge branch 'master' into matrix_spectrum

abentkamp committed:
remove duplicate lemma

abentkamp committed:
fix proof

abentkamp committed:
fix proof
